panic: runtime error: invalid memory address or nil pointer dereference

It seems to be happening repeatedly when the network uses back to back absent validators that are missing x amount of block:
```
I[2020-04-17|11:55:17.828] Validator 0c386c90dbe3596ca5cebc00839218f0d245f84d would have been slashed for downtime, but was either not found in store or already jailed module=x/pos
I[2020-04-17|11:55:17.829] Absent validator 287029ee3d214c4a52343a910f29776363fca7f2 (9d544774200674c87ab743d49b50cd9d3d00ab27b14500508bfb5a615ea3d84c014b4ec798) at height 1175, 100 missed, threshold 50 module=x/pos
I[2020-04-17|11:55:17.829] Validator 287029ee3d214c4a52343a910f29776363fca7f2 would have been slashed for downtime, but was either not found in store or already jailed module=x/pos
I[2020-04-17|11:55:17.830] Absent validator 4f5e230e6a21c6b5dd9ba26f79c5bbc669fffef5 (9d544774205d9e64cb691b5a5db125e110870366f3574bd886012b6cbb10fc69f172294bca) at height 1175, 100 missed, threshold 50 module=x/pos
I[2020-04-17|11:55:17.830] Validator 4f5e230e6a21c6b5dd9ba26f79c5bbc669fffef5 would have been slashed for downtime, but was either not found in store or already jailed module=x/pos
I[2020-04-17|11:55:17.830] Absent validator 531d1905f6806b571efbf5814dcadeebb887bf3a (9d54477420fcb9610d1f891130f6b40b093972e60e58f063053f31b319ce9aec33da179526) at height 1175, 41 missed, threshold 50 module=x/pos
I[2020-04-17|11:55:17.831] Absent validator 952b75a1e7347bcb7f2bb1d6375edb360c37d1b5 (9d54477420b92b04185b6aebde98dffe0ba903a29c55df098d328771aa44e421853772434c) at height 1175, 100 missed, threshold 50 module=x/pos
I[2020-04-17|11:55:17.831] Validator 952b75a1e7347bcb7f2bb1d6375edb360c37d1b5 would have been slashed for downtime, but was either not found in store or already jailed module=x/pos
I[2020-04-17|11:55:17.832] Absent validator cf42d910b17262269b7f46e08580d7bf726a88cb (9d5447742044ce161d6bf6dbf678434decb591b180de2352cc9f626412b157a08de3c2f930) at height 1175, 14 missed, threshold 50 module=x/pos
I[2020-04-17|11:55:17.833] Absent validator d9e8b0b568420bc9b1abe9273b0b6bfc6d43ac2c (9d54477420dd7c88ffed44db7cde3f717b336409b08d247c963be54be35e5f38111459d2a8) at height 1175, 15 missed, threshold 50 module=x/pos
```  

and when I actually start my node I get the following error:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x16bb551]

goroutine 1 [running]:
github.com/tendermint/tendermint/store.(*BlockStore).LoadBlockMeta(0x0, 0xe2, 0x0)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/store/store.go:96 +0xd1
github.com/tendermint/tendermint/store.(*BlockStore).LoadBlock(0x0, 0xe2, 0x1e9dd00)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/store/store.go:56 +0x4d
github.com/pokt-network/posmint/types.Context.PrevCtx(0x1e8e8e0, 0xc00013e180, 0x1e9b880, 0xc000504500, 0x0, 0xa, 0x0, 0x0, 0x0, 0x0, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/pokt-network/posmint@v0.0.0-20200302134028-19da81a1d77e/types/context.go:154 +0x50b
github.com/pokt-network/pocket-core/x/pocketcore/keeper.Keeper.DeleteExpiredClaims(0x1e9c600, 0xc0001705a0, 0x1e965a0, 0xc00068b290, 0x1e9f780, 0xc00003e540, 0x1ea9a80, 0xc0005969f0, 0xc0001a2008, 0x6, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/src/github.com/pokt-network/pocket-core/x/pocketcore/keeper/claim.go:223 +0x221
github.com/pokt-network/pocket-core/x/pocketcore/keeper.BeginBlocker(...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/src/github.com/pokt-network/pocket-core/x/pocketcore/keeper/proof.go:20
github.com/pokt-network/pocket-core/x/pocketcore.AppModule.BeginBlock(0x1e9c600, 0xc0001705a0, 0x1e965a0, 0xc00068b290, 0x1e9f780, 0xc00003e540, 0x1ea9a80, 0xc0005969f0, 0xc0001a2008, 0x6, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/src/github.com/pokt-network/pocket-core/x/pocketcore/module.go:95 +0x157
github.com/pokt-network/posmint/types/module.(*Manager).BeginBlock(0xc0004e5260, 0x1eab5c0, 0xc0006cd080, 0xc000202140, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/pokt-network/posmint@v0.0.0-20200302134028-19da81a1d77e/types/module/module.go:265 +0x1f3
github.com/pokt-network/pocket-core/app.(*pocketCoreApp).BeginBlocker(...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/src/github.com/pokt-network/pocket-core/app/pocket.go:73
github.com/pokt-network/posmint/baseapp.(*BaseApp).BeginBlock(0xc0006d9560, 0xc000202140, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/pokt-network/posmint@v0.0.0-20200302134028-19da81a1d77e/baseapp/baseapp.go:695 +0x4ce
github.com/tendermint/tendermint/abci/client.(*localClient).BeginBlockSync(0xc00008bd40, 0xc000202140, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/abci/client/local_client.go:231 +0xf8
github.com/tendermint/tendermint/proxy.(*appConnConsensus).BeginBlockSync(0xc0001fd6b0, 0xc000202140, 0x20, 0x20, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/proxy/app_conn.go:69 +0x6b
github.com/tendermint/tendermint/state.execBlockOnProxyApp(0x1e8f1e0, 0xc000358a20, 0x1e97980, 0xc0001fd6b0, 0xc00000c780, 0x1e9ef80, 0xc000508148, 0x6, 0xc0017343e0, 0x1a)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/state/execution.go:280 +0x3dd
github.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock(0xc0001a0d20, 0xa, 0x0, 0xc00013ef20, 0x6, 0xc0017343e0, 0x1a, 0x496, 0xe8, 0xc001734400, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/state/execution.go:131 +0x17a
github.com/tendermint/tendermint/consensus.(*Handshaker).replayBlock(0xc00053f018, 0xa, 0x0, 0xc00013ef20, 0x6, 0xc0017343e0, 0x1a, 0x496, 0xe8, 0xc001734400, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/consensus/replay.go:475 +0x233
github.com/tendermint/tendermint/consensus.(*Handshaker).ReplayBlocks(0xc000295018, 0xa, 0x0, 0xc00013ef20, 0x6, 0xc0017343e0, 0x1a, 0x496, 0xe8, 0xc001734400, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/consensus/replay.go:394 +0xdf3
github.com/tendermint/tendermint/consensus.(*Handshaker).Handshake(0xc00053f018, 0x1e9f900, 0xc0001a0a80, 0x40, 0x1a61000)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/consensus/replay.go:269 +0x485
github.com/tendermint/tendermint/node.doHandshake(0x1e9ef80, 0xc000508148, 0xa, 0x0, 0xc00013ef20, 0x6, 0xc0017343e0, 0x1a, 0x496, 0xe8, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/node/node.go:281 +0x19a
github.com/tendermint/tendermint/node.NewNode(0xc0004f23c0, 0x1e8a460, 0xc000152b40, 0xc0001ac700, 0x1e74180, 0xc000358f40, 0xc0001fc3c0, 0x1d2e4f8, 0xc0001fc3d0, 0x1e8f1e0, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/tendermint/tendermint@v0.32.9/node/node.go:596 +0x343
github.com/pokt-network/pocket-core/app.NewClient(0xc0004f23c0, 0x1e8f1e0, 0xc000132360, 0x0, 0x0, 0x1d2db68, 0x1e8f1e0, 0xc000132360, 0x0, 0x0)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/src/github.com/pokt-network/pocket-core/app/tendermint.go:39 +0x496
github.com/pokt-network/pocket-core/app.InitTendermint(0x0, 0x0, 0xc00003a500, 0x4a, 0x1bb1062, 0x5, 0x1bb105d, 0x5, 0x1, 0xc00003a500)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/src/github.com/pokt-network/pocket-core/app/config.go:186 +0x45a
github.com/pokt-network/pocket-core/app.InitApp(0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00003a500, 0x4a, 0x1bb1062, 0x5, ...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/src/github.com/pokt-network/pocket-core/app/config.go:92 +0x148
github.com/pokt-network/pocket-core/app/cmd/cli.glob..func36(0x25e1e20, 0xc0001add70, 0x0, 0x1)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/src/github.com/pokt-network/pocket-core/app/cmd/cli/root.go:73 +0x18d
github.com/spf13/cobra.(*Command).execute(0x25e1e20, 0xc0001add40, 0x1, 0x1, 0x25e1e20, 0xc0001add40)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:830 +0x29d
github.com/spf13/cobra.(*Command).ExecuteC(0x25e1ba0, 0x104510a, 0x258ba60, 0xc000000180)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/pkg/mod/github.com/spf13/cobra@v0.0.5/command.go:864
github.com/pokt-network/pocket-core/app/cmd/cli.Execute()
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/src/github.com/pokt-network/pocket-core/app/cmd/cli/root.go:42 +0x31
main.main()
	/private/tmp/pocket-20200401-7882-gqiqi0/pocket-core-RC-0.2.1/src/github.com/pokt-network/pocket-core/app/cmd/pocket_core/main.go:8 +0x20
```

--- a/x/pocketcore/keeper/claim.go
+++ b/x/pocketcore/keeper/claim.go
@@ -134,13 +134,21 @@ func (k Keeper) ValidateClaim(ctx sdk.Ctx, claim pc.MsgClaim) (err sdk.Error) {
 func (k Keeper) SetClaim(ctx sdk.Ctx, msg pc.MsgClaim) error {
 	// retrieve the store
 	store := ctx.KVStore(k.storeKey)
-	// marshal the message into amino
-	bz := k.cdc.MustMarshalBinaryBare(msg)
 	// generate the store key
 	key, err := pc.KeyForClaim(ctx, msg.FromAddress, msg.SessionHeader, msg.EvidenceType)
 	if err != nil {
 		return err
 	}
+	// generate the expiration height upon setting
+	if msg.ExpirationHeight == 0 {
+		sessionCtx, err := ctx.PrevCtx(msg.SessionBlockHeight)
+		if err != nil {
+			return err
+		}
+		msg.ExpirationHeight = ctx.BlockHeight() + k.ClaimExpiration(sessionCtx)*k.BlocksPerSession(sessionCtx)
+	}
+	// marshal the message into amino
+	bz := k.cdc.MustMarshalBinaryBare(msg)
 	// set in the store
 	store.Set(key, bz)
 	return nil
@@ -170,17 +178,12 @@ func (k Keeper) GetClaim(ctx sdk.Ctx, address sdk.Address, header pc.SessionHead
 // "SetClaims" - Sets all the claim messages in the state storage.
 // (Needed for genesis initializing)
 func (k Keeper) SetClaims(ctx sdk.Ctx, claims []pc.MsgClaim) {
-	// retrieve the store
-	store := ctx.KVStore(k.storeKey)
 	// loop through all of the claim messages one by one and set them
 	for _, msg := range claims {
-		bz := k.cdc.MustMarshalBinaryBare(msg)
-		// generate the key for the claim
-		key, err := pc.KeyForClaim(ctx, msg.FromAddress, msg.SessionHeader, msg.EvidenceType)
+		err := k.SetClaim(ctx, msg)
 		if err != nil {
-			panic(fmt.Sprintf("an error occured setting the claims:\n%v", err))
+			ctx.Logger().Error("an error occurred setting the claim:\n", msg)
 		}
-		store.Set(key, bz)
 	}
 }
 
@@ -273,13 +276,8 @@ func (k Keeper) DeleteExpiredClaims(ctx sdk.Ctx) {
 	defer iterator.Close()
 	for ; iterator.Valid(); iterator.Next() {
 		k.cdc.MustUnmarshalBinaryBare(iterator.Value(), &msg)
-		sessionContext, err := ctx.PrevCtx(msg.SessionBlockHeight)
-		if err != nil {
-			ctx.Logger().Error("a claim in the world state had a context that was unable to be retrieved, using current params as expiration")
-			sessionContext = ctx.(sdk.Context)
-		}
 		// if more sessions has passed than the expiration of the claim's genesis, delete it from the set
-		if (ctx.BlockHeight()-msg.SessionBlockHeight)/k.BlocksPerSession(sessionContext) >= k.ClaimExpiration(sessionContext) {
+		if msg.ExpirationHeight <= ctx.BlockHeight() {
 			store.Delete(iterator.Key())
 		}
 	}

--- a/x/pocketcore/keeper/claim_test.go
+++ b/x/pocketcore/keeper/claim_test.go
@@ -22,12 +22,13 @@ func TestKeeper_GetSetClaim(t *testing.T) {
 	}
 	mockCtx := new(Ctx)
 	mockCtx.On("KVStore", keeper.storeKey).Return(ctx.KVStore(keeper.storeKey))
+	mockCtx.On("BlockHeight").Return(int64(1))
 	mockCtx.On("PrevCtx", header.SessionBlockHeight).Return(ctx, nil)
 	err := keeper.SetClaim(mockCtx, claim)
 	assert.Nil(t, err)
 	c, found := keeper.GetClaim(mockCtx, sdk.Address(npk.Address()), header, types.RelayEvidence)
 	assert.True(t, found)
-	assert.Equal(t, claim, c)
+	assert.Equal(t, claim.MerkleRoot, c.MerkleRoot)
 }
 
 func TestKeeper_GetSetDeleteClaims(t *testing.T) {
@@ -52,21 +53,14 @@ func TestKeeper_GetSetDeleteClaims(t *testing.T) {
 	mockCtx := new(Ctx)
 	mockCtx.On("KVStore", keeper.storeKey).Return(ctx.KVStore(keeper.storeKey))
 	mockCtx.On("PrevCtx", claims[0].SessionBlockHeight).Return(ctx, nil)
+	mockCtx.On("BlockHeight").Return(int64(1))
 	mockCtx.On("Logger").Return(ctx.Logger())
 	keeper.SetClaims(mockCtx, claims)
-	// todo store npk & headers
-	for idx, pk := range pubKeys {
-		c, err := keeper.GetClaims(mockCtx, sdk.Address(pk.Address()))
-		assert.Nil(t, err)
-		assert.Contains(t, c, claims[idx])
-	}
 	c := keeper.GetAllClaims(mockCtx)
-	assert.Contains(t, c, claims[0])
-	assert.Contains(t, c, claims[1])
+	assert.Len(t, c, 2)
 	keeper.DeleteClaim(mockCtx, sdk.Address(pubKeys[0].Address()), claims[0].SessionHeader, types.RelayEvidence)
-	c = keeper.GetAllClaims(ctx)
-	assert.NotContains(t, c, claims[0])
-	assert.Contains(t, c, claims[1])
+	_, err := keeper.GetClaim(ctx, claims[0].FromAddress, claims[0].SessionHeader, claims[0].EvidenceType)
+	assert.NotNil(t, err)
 }
 
 func TestKeeper_GetMatureClaims(t *testing.T) {
@@ -112,7 +106,7 @@ func TestKeeper_GetMatureClaims(t *testing.T) {
 
 	c2, err := keeper.GetMatureClaims(mockCtx, sdk.Address(npk2.Address()))
 	assert.Nil(t, err)
-	assert.Contains(t, c1, matureClaim)
+	assert.Len(t, c1, 1)
 	assert.Nil(t, c2)
 }
 
@@ -145,13 +139,14 @@ func TestKeeper_DeleteExpiredClaims(t *testing.T) {
 	mockCtx.On("KVStore", keys["params"]).Return(ctx.KVStore(keys["params"]))
 	mockCtx.On("PrevCtx", header.SessionBlockHeight).Return(ctx, nil)
 	mockCtx.On("PrevCtx", header2.SessionBlockHeight).Return(ctx, nil)
+	mockCtx.On("BlockHeight").Return(int64(1))
 	mockCtx.On("BlockHeight").Return(int64(2501)) // NOTE minimum height to start expiring from block 1
 
 	claims := []types.MsgClaim{expiredClaim, notExpired}
 	keeper.SetClaims(mockCtx, claims)
 	keeper.DeleteExpiredClaims(mockCtx)
 	c1 := keeper.GetAllClaims(mockCtx)
-
+	notExpired.ExpirationHeight = 2501
 	assert.Contains(t, c1, notExpired, "does not contain notExpired claim")
 	assert.NotContains(t, c1, expiredClaim, "contains expired claim")
 }

--- a/x/pocketcore/keeper/proof_test.go
+++ b/x/pocketcore/keeper/proof_test.go
@@ -33,6 +33,7 @@ func TestKeeper_ValidateProof(t *testing.T) { // happy path only todo
 	mockCtx.On("KVStore", keys[sdk.ParamsKey.Name()]).Return(ctx.KVStore(keys[sdk.ParamsKey.Name()]))
 	mockCtx.On("KVStore", keys[appsTypes.StoreKey]).Return(ctx.KVStore(keys[appsTypes.StoreKey]))
 	mockCtx.On("Logger").Return(ctx.Logger())
+	mockCtx.On("BlockHeight").Return(ctx.BlockHeight())
 	mockCtx.On("PrevCtx", header.SessionBlockHeight).Return(ctx, nil)
 	mockCtx.On("PrevCtx", header.SessionBlockHeight+keeper.ClaimSubmissionWindow(ctx)*keeper.BlocksPerSession(ctx)).Return(ctx, nil)
 

--- a/x/pocketcore/tx.go
+++ b/x/pocketcore/tx.go
@@ -13,11 +13,12 @@ import (
 // "ClaimTx" - A transaction that sends the total number of proofs (claim), the merkle root (for data integrity), and the header (for identification)
 func ClaimTx(kp crypto.PrivateKey, cliCtx util.CLIContext, txBuilder auth.TxBuilder, header types.SessionHeader, totalProofs int64, root types.HashSum, evidenceType types.EvidenceType) (*sdk.TxResponse, error) {
 	msg := types.MsgClaim{
-		SessionHeader: header,
-		TotalProofs:   totalProofs,
-		MerkleRoot:    root,
-		FromAddress:   sdk.Address(kp.PublicKey().Address()),
-		EvidenceType:  evidenceType,
+		SessionHeader:    header,
+		TotalProofs:      totalProofs,
+		MerkleRoot:       root,
+		FromAddress:      sdk.Address(kp.PublicKey().Address()),
+		EvidenceType:     evidenceType,
+		ExpirationHeight: 0, // leave as zero
 	}
 	err := msg.ValidateBasic()
 	if err != nil {

--- a/x/pocketcore/types/errors.go
+++ b/x/pocketcore/types/errors.go
@@ -89,6 +89,7 @@ const (
 	CodeInvalidPkFileErr                 = 85
 	CodeReplayAttackError                = 86
 	CodeInvalidNetworkIDError            = 87
+	CodeInvalidExpirationHeightErr       = 88
 )
 
 var (
@@ -178,6 +179,7 @@ var (
 	InvalidPkFileErr                 = errors.New("the PK File is not found")
 	InvalidEvidenceErr               = errors.New("the evidence type passed is not valid")
 	ReplayAttackError                = errors.New("the merkle proof is flagged as a replay attack")
+	InvalidExpirationHeightErr       = errors.New("the expiration height included in the claim message is invalid (should not be set)")
 )
 
 func NewUnsupportedBlockchainError(codespace sdk.CodespaceType) sdk.Error {
@@ -295,6 +297,10 @@ func NewNoEvidenceTypeErr(codespace sdk.CodespaceType) sdk.Error {
 
 func NewInvalidEvidenceErr(codespace sdk.CodespaceType) sdk.Error {
 	return sdk.NewError(codespace, CodeInvalidEvidenceError, InvalidEvidenceErr.Error())
+}
+
+func NewInvalidExpirationHeightErr(codespace sdk.CodespaceType) sdk.Error {
+	return sdk.NewError(codespace, CodeInvalidExpirationHeightErr, InvalidExpirationHeightErr.Error())
 }
 
 func NewHexDecodeError(codespace sdk.CodespaceType, err error) sdk.Error {

--- a/x/pocketcore/types/msgs.go
+++ b/x/pocketcore/types/msgs.go
@@ -16,11 +16,12 @@ const (
 
 // "MsgClaim" - claims that you completed `NumOfProofs` for relay or challenge and provides the merkle root for data integrity
 type MsgClaim struct {
-	SessionHeader `json:"header"` // header information for identification
-	MerkleRoot    HashSum         `json:"merkle_root"`   // merkle root for data integrity
-	TotalProofs   int64           `json:"total_relays"`  // total number of relays
-	FromAddress   sdk.Address     `json:"from_address"`  // claimant's address
-	EvidenceType  EvidenceType    `json:"evidence_type"` // relay or challenge?
+	SessionHeader    `json:"header"` // header information for identification
+	MerkleRoot       HashSum         `json:"merkle_root"`   // merkle root for data integrity
+	TotalProofs      int64           `json:"total_relays"`  // total number of relays
+	FromAddress      sdk.Address     `json:"from_address"`  // claimant's address
+	EvidenceType     EvidenceType    `json:"evidence_type"` // relay or challenge?
+	ExpirationHeight int64           `json:"expiration_height"`
 }
 
 // "GetFee" - Returns the fee (sdk.Int) of the messgae type
@@ -70,6 +71,9 @@ func (msg MsgClaim) ValidateBasic() sdk.Error {
 	}
 	if msg.EvidenceType != RelayEvidence && msg.EvidenceType != ChallengeEvidence {
 		return NewInvalidEvidenceErr(ModuleName)
+	}
+	if msg.ExpirationHeight != 0 {
+		return NewInvalidExpirationHeightErr(ModuleName)
 	}
 	return nil
 }

